File: benchmark/confusions_recitation.py
"""Les confusions de recitation, et leur rendement MESURE. Source unique.

Ce module ne depend de rien : il est importe aussi bien par les generateurs
(venv TTS) que par les bancs (venv NeMo). La table vivait dans
`generate_phrases_concat.py`, qui charge XTTS a l'import -- inutilisable depuis
un banc. La dupliquer aurait garanti qu'elle diverge : c'est une table de
MESURES, pas une constante de gout.

RENDEMENT = part des clips ou la faute S'ENTEND reellement, mesuree sur mot
isole par rapport de vraisemblance CTC (controle du 2026-07-31, 18 195 clips du
corpus `tts_augmentation`).

⚠️ Ce rendement vaut pour un MOT ISOLE. En synthese de PHRASE entiere, XTTS
corrige le texte qu'on lui donne exactement comme l'ASR : 29 % a 2-3 mots, 10 %
a 3-6 mots. C'est ce qui a impose l'assemblage mot a mot.
"""

# Vraies confusions de recitation : meme point d'articulation, l'emphase seule
# change. Ce sont celles qu'on cherche a detecter, et XTTS les rend tres bien
# tant qu'il n'a pas de contexte a corriger.
CONFUSABLES = [
    ("ط", "ت"),   # 99 %
    ("ت", "ط"),   # 99 %
    ("ض", "د"),   # 99 %
    ("د", "ض"),   # 99 %
    ("ق", "ك"),   # 98 %
    ("ك", "ق"),   # 98 %
    ("ص", "س"),   # 98 %
    ("ه", "ح"),   # 97 %
    ("ح", "ه"),   # 97 %
    ("ء", "ع"),   # 97 %
    ("ذ", "ز"),   # 97 %
    ("س", "ص"),   # 96 %
    ("ع", "ء"),   # 95 %
    # Famille jim/ha/kha : ABSENTE du corpus d'origine, signalee par
    # l'utilisateur apres son propre test (« mon cas c'etait remplacer jim par
    # ha »). Rendement inconnu -- c'est le controle qui tranchera.
    ("ج", "ح"), ("ح", "خ"), ("ج", "خ"),
]

# NE PAS SUPPRIMER : seule substitution du corpus d'origine que XTTS ne rend
# jamais en phrase (0/4), alors que ز->ذ passe a 97 % sur mot isole. Exclue de
# la GENERATION tant que ce n'est pas explique -- mais elle reste une confusion
# legitime, donc elle demeure dans la table pour les BANCS.
EXCLUES_GENERATION = {"ز->ذ"}

FATHA, DAMMA, KASRA, SUKUN = "َ", "ُ", "ِ", "ْ"
SHORT_HARAKAT = [FATHA, DAMMA, KASRA, SUKUN]


def variantes(mot):
    """Toutes les confusions plausibles de ce mot -- une lettre, ou une harakat.

    Sert au banc des regles de gop : comparer le mot attendu a sa meilleure
    confusion REELLE, plutot qu'a un decodage libre qui est un max sur tout le
    lexique et donc une borne tres lache.
    """
    out = []
    for a, b in CONFUSABLES:
        for src, dst in ((a, b), (b, a)):
            j = mot.find(src)
            if j >= 0:
                out.append(mot[:j] + dst + mot[j + 1:])
    for k, c in enumerate(mot):
        if c in SHORT_HARAKAT:
            for h in SHORT_HARAKAT:
                if h != c:
                    out.append(mot[:k] + h + mot[k + 1:])
    return list(dict.fromkeys(out))

File: app/android/app/src/main/kotlin/com/corankarim/coran_karim/recitation2/ConfusionsRecitation.kt
package com.corankarim.coran_karim.recitation2

/**
 * LES CONFUSIONS TELLES QUE LE PYTHON LES GENERE — port de
 * `benchmark/confusions_recitation.py::variantes`.
 *
 * ── POURQUOI UN SECOND GENERATEUR DE VARIANTES DANS CE PROJET ──────────────
 *
 * `fastconformer/ConfusableVariants.kt` existe deja et sert au rescoring de la
 * chaine v1. Il N'EST PAS interchangeable avec celui-ci, et les confondre
 * casserait la tete 3 sans que rien ne le signale. Trois divergences mesurees
 * en comparant les deux inventaires :
 *
 *   |                     | ConfusableVariants        | ici (= Python)          |
 *   |---------------------|---------------------------|-------------------------|
 *   | position substituee | TOUTES les occurrences    | la PREMIERE seulement   |
 *   | harakat             | 7 (avec les 3 tanwins)    | 4 (fatha/damma/kasra/sukun) |
 *   | paires de lettres   | table symetrisee a la main| la liste CONFUSABLES     |
 *
 * Chacune change le jeu de candidats, donc `alt` et `alt2`, donc deux des douze
 * caracteristiques de la tete. Une tete entrainee sur un jeu et interrogee sur
 * un autre ne rend pas une erreur : elle rend un nombre faux.
 *
 * ⇒ Celui-ci sert EXCLUSIVEMENT a la tete 3. `ConfusableVariants` reste la
 * source du rescoring v1, dont le rendement (80,8 % sur les lettres) a ete
 * mesure avec SA table a lui. Ne pas unifier les deux sans re-mesurer les deux.
 *
 * ── L'ORDRE DE LA TABLE EST CELUI DU PYTHON, PAS UN CHOIX ──────────────────
 *
 * La deduplication garde la PREMIERE apparition (`dict.fromkeys` cote Python,
 * LinkedHashSet ici). Reordonner la table changerait donc la liste produite sur
 * les mots ou deux regles engendrent la meme variante.
 */
object ConfusionsRecitation {

    /**
     * Paires de confusion, dans l'ordre exact de `CONFUSABLES`. Le pourcentage
     * est la part des clips ou la faute S'ENTEND reellement (18 195 clips,
     * controle du 2026-07-31) -- c'est une table de MESURES, pas de gout.
     *
     * ⚠️ La liste contient deja les deux sens pour la plupart des paires
     * (ط→ت ET ت→ط) et le generateur essaie DE TOUTE FACON les deux sens de
     * chaque entree. Ce doublon est dans le Python ; il est sans effet grace a
     * la deduplication, mais il faut le reproduire pour que l'ORDRE des
     * variantes reste le meme.
     */
    private val CONFUSABLES: List<Pair<Char, Char>> = listOf(
        'ط' to 'ت',   // 99 %
        'ت' to 'ط',   // 99 %
        'ض' to 'د',   // 99 %
        'د' to 'ض',   // 99 %
        'ق' to 'ك',   // 98 %
        'ك' to 'ق',   // 98 %
        'ص' to 'س',   // 98 %
        'ه' to 'ح',   // 97 %
        'ح' to 'ه',   // 97 %
        'ء' to 'ع',   // 97 %
        'ذ' to 'ز',   // 97 %
        'س' to 'ص',   // 96 %
        'ع' to 'ء',   // 95 %
        // Famille jim/ha/kha : ABSENTE du corpus d'origine, signalee par
        // l'utilisateur apres son propre test. Rendement inconnu.
        'ج' to 'ح', 'ح' to 'خ', 'ج' to 'خ',
    )

    /** QUATRE harakat, pas sept : le Python n'inclut PAS les tanwins ici. */
    private val HARAKAT_COURTES = charArrayOf('َ', 'ُ', 'ِ', 'ْ')

    /**
     * Toutes les confusions plausibles du mot : une lettre, ou une harakat.
     *
     * ⚠️ SUBSTITUTION DE LA PREMIERE OCCURRENCE SEULEMENT pour les lettres --
     * c'est `mot.find(src)` cote Python. Substituer toutes les occurrences
     * produirait des variantes que la tete n'a jamais vues.
     */
    fun variantes(mot: String): List<String> {
        val out = LinkedHashSet<String>()
        for ((a, b) in CONFUSABLES) {
            for ((src, dst) in listOf(a to b, b to a)) {
                val j = mot.indexOf(src)
                if (j >= 0) out.add(mot.substring(0, j) + dst + mot.substring(j + 1))
            }
        }
        for (k in mot.indices) {
            val c = mot[k]
            if (c in HARAKAT_COURTES) {
                for (h in HARAKAT_COURTES) {
                    if (h != c) out.add(mot.substring(0, k) + h + mot.substring(k + 1))
                }
            }
        }
        return out.toList()
    }
}
